File: 46. 全排列/swift/main/main/main.swift
//
//  main.swift
//  main
//
//  Created by ByteDance on 4/28/25.
//

import Foundation

class Solution {
    var result = [[Int]]()
    var path = [Int]()
    var used = [Int: Bool]()

    func permute(_ nums: [Int]) -> [[Int]] {
        backtracking(nums)
        return result
    }

    func backtracking(_ nums: [Int]) {
        if path.count == nums.count {
            result.append(path)
            return
        }

        for i in 0 ..< nums.count {
            let num = nums[i]
            if used[num] == true {
                continue
            }
            path.append(num)
            used[num] = true
            backtracking(nums)
            used[num] = false
            path.removeLast()
        }
    }
}

File: 46. 全排列/go/main.go
package main

import "fmt"

func main() {
	a := []int{1, 2, 3, 4}

	fmt.Println(permute(a))
}

func permute(nums []int) [][]int {
	if len(nums) == 1 {
		return [][]int{nums}
	}

	var result = [][]int{}

	for i, num := range nums {
		combine := make([]int, len(nums)-1)
		copy(combine[0:], nums[0:i])
		copy(combine[i:], nums[i+1:])

		subResult := permute(combine)
		for _, s := range subResult {
			result = append(result, append(s, num))
		}
	}

	return result
}
